### PR TITLE
Refactor iteration over nested collection elements

### PR DESCRIPTION
**Pull request by joshessman-llnl**
# Summary

- This PR is a refactoring
- It does the following:
  - Moves the duplicated iteration logic in the `inlet::Container` implementation to a single function that accepts a lambda

This should prevent bugs like #467, where changes to iteration logic are not propagated to all the places that need it.  Instead, only the unique logic is present at each call site.

### CODE REVIEW

**Review by white238 — APPROVED** [review]

File: src/axom/inlet/Container.hpp [hunk 1]
```diff
@@ -1231,6 +1231,34 @@ class Container : public Verifiable<Container>
   template <typename Func>
   void forEachCollectionElement(Func&& func) const;
 
+  /*!
+   *****************************************************************************
+   * \brief Applies a provided function to nested elements of the calling table
+   * and stores the result in a range
+   * 
+   * \pre The function must accept two arguments of type Table& and
```
Comment (kennyweiss) [inline src/axom/inlet/Container.hpp:1239]:
```suggestion
   * \pre The function \a func must accept two arguments of type Table& and
```

File: src/axom/inlet/Container.hpp [hunk 2]
```diff
@@ -1231,6 +1231,34 @@ class Container : public Verifiable<Container>
   template <typename Func>
   void forEachCollectionElement(Func&& func) const;
 
+  /*!
+   *****************************************************************************
+   * \brief Applies a provided function to nested elements of the calling table
+   * and stores the result in a range
```
Comment (kennyweiss) [inline src/axom/inlet/Container.hpp:1237]: On my first pas through the docs, I am not sure what "...in a range" means at the end

File: src/axom/inlet/Container.hpp [hunk 3]
```diff
@@ -1231,6 +1231,34 @@ class Container : public Verifiable<Container>
...
+   * 
+   * \pre The function must accept two arguments of type Table& and
+   * const std::string&, respectively.  
+   * 
+   * This function will pass to the provided function the nested table
+   * as the first argument and the path of the nested element in the input file
+   * as the second argument, when applicable.
+   * 
+   * \param [out] output An iterator to the beginning of the output range
```
Comment (kennyweiss) [inline src/axom/inlet/Container.hpp:1246]: Minor / convention -- is it weird for the first parameter to be an output parameter, while the others are input parameters? Would users expect the output parameters to be at the end?

File: src/axom/inlet/Container.hpp (hunk 2 above)
Comment (kennyweiss) [reply src/axom/inlet/Container.hpp:1237]: Update after seeing the rest of the docs -- the results are put into an output iterator ``\a output``

File: src/axom/inlet/Container.hpp (hunk 1 above)
Comment (kennyweiss) [reply src/axom/inlet/Container.hpp:1239]: Also, can you please check that doxygen doesn't choke on the ampersand in `Table&`

**Review by kennyweiss — APPROVED** [review]
Thanks @joshessman-llnl 

This cleans up the code a bit, but also makes its internals more complex to use.

Thinking a bit about future maintainers, I think it would make sense to add a test for this function to help people see (in isolation) how it is meant to be used.

File: src/axom/inlet/Container.hpp (hunk 3 above)
Comment (joshessman-llnl) [reply src/axom/inlet/Container.hpp:1259]: Good question - STL algorithms like [std::transform](https://en.cppreference.com/w/cpp/algorithm/transform) typically always keep the callable at the end for readability when passing a lambda as an argument, and that's what I've opted to go with there.  I don't have a preference regarding the order of the first two, though, so I can flip those if you'd prefer.

Comment (kennyweiss) [reply src/axom/inlet/Container.hpp:1259]: That makes sense.  I'm ok with it as is.

File: src/axom/inlet/Container.hpp (hunk 1 above)
Comment (joshessman-llnl) [reply src/axom/inlet/Container.hpp:1252]: I think Doxygen doesn't render private methods so this would only be for developer use

**Comment by joshessman-llnl** [comment]
> 
> Thinking a bit about future maintainers, I think it would make sense to add a test for this function to help people see (in isolation) how it is meant to be used.

This might be a little difficult to implement given that it's a private method - there are of course workarounds but I'm not sure what the tradeoff is.

**Comment by white238** [comment]
> > Thinking a bit about future maintainers, I think it would make sense to add a test for this function to help people see (in isolation) how it is meant to be used.
> 
> This might be a little difficult to implement given that it's a private method - there are of course workarounds but I'm not sure what the tradeoff is.

This is being tested at a high level but would be hard to test this specifically at a unit test level.  I think this is fine to merge.

**Pull request merged**